test/test_subclass.py

```diff
@@ -0,0 +1,206 @@
+            if as_param:
+                # Serialization should preserve both custom type and "parameter-ness".
+                self.assertTrue(isinstance(x_loaded, tensor_cls))
+                self.assertTrue(isinstance(x_loaded, nn.Parameter))
+            else:
+                self.assertTrue(type(x_loaded) is tensor_cls)
+
+    @parametrize_tensor_cls
+    @parametrize("as_param", [False, True])
+    def test_repr(self, tensor_cls, as_param):
+        x = self._create_tensor(tensor_cls)
+        if as_param:
+            x = nn.Parameter(x)
+        str_repr = x.__repr__()
+        if tensor_cls is not torch.Tensor:
+            self.assertEqual(str_repr.count(f"{tensor_cls.__name__}("), 1)
+        self.assertEqual(str_repr.count("Parameter"), 1 if as_param else 0)
+
+    @parametrize_tensor_cls
+    @parametrize("as_param", [False, True])
+    def test_type_propagation(self, tensor_cls, as_param):
+        x = self._create_tensor(tensor_cls)
+        if as_param:
+            x = nn.Parameter(x)
+
+        # Call the add operator to produce an output tensor.
+        output = x + self._create_tensor(torch.Tensor)
+
+        # Custom type should be propagated across operations, but "parameter-ness" should not be.
+        self.assertTrue(output.__class__ is (tensor_cls if as_param else x.__class__))
+
+    @parametrize_tensor_cls
+    def test_module_optimization(self, tensor_cls):
+        def create_fn():
+            return self._create_tensor(tensor_cls)
+
+        class MyModule(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.p1 = nn.Parameter(create_fn())
+
+                self.p_list = nn.ParameterList([create_fn() for _ in range(3)])
+                self.p_list.append(create_fn())
+
+                self.p_dict = nn.ParameterDict({
+                    'foo': create_fn(),
+                    'bar': create_fn(),
+                })
+                self.p_dict['baz'] = create_fn()
+
+                with torch.no_grad():
+                    nn.init.normal_(self.p1)
+                    for p in self.p_list:
+                        nn.init.uniform_(p)
+                    for _, p in self.p_dict.items():
+                        nn.init.uniform_(p)
+
+            def forward(self, x):
+                out = self.p1 + x
+                for p in self.p_list:
+                    out = p + out
+
+                for _, v in self.p_dict.items():
+                    out = v + out
+
+                return out
+
+        m = MyModule()
+        self.assertEqual(len(m.state_dict()), 8)
+
+        optimizer = torch.optim.SGD(m.parameters(), lr=0.1)
+        m(create_fn()).sum().backward(torch.tensor(1))
+        optimizer.step()
+
+    @parametrize("tensor_cls", [
+        subtest(tensor_cls, name=info.name) for tensor_cls, info in subclass_db.items()
+        # TODO: Either implement set_() properly for these tensor subclasses or apply a
```

Comment:
nit: have these skipped inside the test so that you can use @parametrize_tensor_cls here?